pkg/services/dashboards/database/database_dashboard_public.go

```diff
@@ -0,0 +1,157 @@
+	if uid == "" {
+		return nil, nil, models.ErrPublicDashboardIdentifierNotSet
+	}
+
+	// get public dashboard
+	pdRes := &models.PublicDashboard{Uid: uid}
+	err := d.sqlStore.WithTransactionalDbSession(context.Background(), func(sess *sqlstore.DBSession) error {
+		has, err := sess.Get(pdRes)
+		if err != nil {
+			return err
+		}
+		if !has {
+			return models.ErrPublicDashboardNotFound
+		}
+		return nil
+	})
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// find dashboard
+	dashRes := &models.Dashboard{OrgId: pdRes.OrgId, Uid: pdRes.DashboardUid}
+	err = d.sqlStore.WithTransactionalDbSession(context.Background(), func(sess *sqlstore.DBSession) error {
+		has, err := sess.Get(dashRes)
+		if err != nil {
+			return err
+		}
+		if !has {
+			return models.ErrPublicDashboardNotFound
+		}
+		return nil
+	})
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return pdRes, dashRes, err
+}
+
+// generates a new unique uid to retrieve a public dashboard
+func generateNewPublicDashboardUid(sess *sqlstore.DBSession) (string, error) {
+	for i := 0; i < 3; i++ {
+		uid := util.GenerateShortUID()
+
+		exists, err := sess.Get(&models.PublicDashboard{Uid: uid})
+		if err != nil {
+			return "", err
+		}
+
+		if !exists {
+			return uid, nil
+		}
+	}
+
+	return "", models.ErrPublicDashboardFailedGenerateUniqueUid
+}
+
+// retrieves public dashboard configuration
+func (d *DashboardStore) GetPublicDashboardConfig(orgId int64, dashboardUid string) (*models.PublicDashboardConfig, error) {
+	if dashboardUid == "" {
+		return nil, models.ErrDashboardIdentifierNotSet
+	}
+
+	// get dashboard and publicDashboard
+	dashRes := &models.Dashboard{OrgId: orgId, Uid: dashboardUid}
+	pdRes := &models.PublicDashboard{OrgId: orgId, DashboardUid: dashboardUid}
+	err := d.sqlStore.WithTransactionalDbSession(context.Background(), func(sess *sqlstore.DBSession) error {
+		// dashboard
+		has, err := sess.Get(dashRes)
+		if err != nil {
+			return err
+		}
+		if !has {
+			return models.ErrDashboardNotFound
+		}
+
+		// publicDashboard
+		_, err = sess.Get(pdRes)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	pdc := &models.PublicDashboardConfig{
+		IsPublic:        dashRes.IsPublic,
+		PublicDashboard: *pdRes,
+	}
+
+	return pdc, err
+}
+
+// persists public dashboard configuration
+func (d *DashboardStore) SavePublicDashboardConfig(cmd models.SavePublicDashboardConfigCommand) (*models.PublicDashboardConfig, error) {
+	if len(cmd.PublicDashboardConfig.PublicDashboard.DashboardUid) == 0 {
+		return nil, models.ErrDashboardIdentifierNotSet
+	}
+
+	err := d.sqlStore.WithTransactionalDbSession(context.Background(), func(sess *sqlstore.DBSession) error {
+		// update isPublic on dashboard entry
+		affectedRowCount, err := sess.Table("dashboard").Where("org_id = ? AND uid = ?", cmd.OrgId, cmd.DashboardUid).Update(map[string]interface{}{"is_public": cmd.PublicDashboardConfig.IsPublic})
```

Comment:
This is problematic, to update the main dashboard table from this function like this. Updates to the dashboard should be via the main update dashboard service, and it should cause an update to version & updated_by updated date, all are sidestepped here. This feels like another indication that this flag should not be in the dashboard table / main dashboard entity but in the config table.

Reply:
We need to ship this to keep moving forward, but I will begin working on a PR Friday to move the isPublic flag.